control-plane/cni/main.go

```diff
@@ -135,56 +162,119 @@ func cmdAdd(args *skel.CmdArgs) error {
+	if err != nil {
+		logger.Error("could not parse annotation: %s, error: %v", annotationCNIProxyConfig, err)
+		return types.PrintResult(result, cfg.CNIVersion)
+	}
+
+	// get the DNS IP from the environment.
+	dnsIP := searchDNSIPFromEnvironment(*pod, cfg.DNSPrefix)
+	if dnsIP != "" {
+		logger.Error("assigned consul dns ip to %s", dnsIP)
+		iptablesCfg.ConsulDNSIP = dnsIP
 	}
 
-	// TODO: Get transparent proxy annotations and merge with proxy config
-	// TODO: Redirect traffic :)
+	// apply the iptables rules.
+	err = iptables.Setup(iptablesCfg)
+	if err != nil {
+		logger.Error("could not apply iptables setup: %v", err)
+		return types.PrintResult(result, cfg.CNIVersion)
+	}
 
-	// Pass through the result for the next plugin
+	logger.Debug("traffic redirect rules applied to pod: %s", pod.Name)
+	// Pass through the result for the next plugin even though we are the final plugin in the chain.
 	return types.PrintResult(result, cfg.CNIVersion)
 }
 
 // cmdDel is called for DELETE requests.
 func cmdDel(args *skel.CmdArgs) error {
-	// Nothing for consul-cni plugin to do as everything is removed once the pod is gone
+	// Nothing to do but this function will still be called as part of the CNI specification.
 	return nil
 }
 
+// cmdCheck is called for CHECK requests.
 func cmdCheck(args *skel.CmdArgs) error {
+	// Nothing to do but this function will still be called as part of the CNI specification.
 	return nil
 }
 
 func main() {
 	skel.PluginMain(cmdAdd, cmdCheck, cmdDel, version.All, bv.BuildString("consul-cni"))
 }
 
-func hasBeenInjected(pod corev1.Pod) bool {
-	if anno, ok := pod.Annotations[keyInjectStatus]; ok && anno == injected {
+// skipConnectInject determines if the connect-inject annotation is false.
+func skipConnectInject(pod corev1.Pod) bool {
+	if anno, ok := pod.Annotations[annotationInject]; ok && anno == "false" {
 		return true
 	}
 	return false
 }
+
+// waitForAnnotation waits for an annotation to be available. Returns immediately if the annotation exists.
+func waitForAnnotation(pod corev1.Pod, annotation string, retries uint64) bool {
+	var err error
+	err = backoff.Retry(func() error {
+		var ok bool
+		_, ok = pod.Annotations[annotation]
+		if !ok {
+			return fmt.Errorf("annotation %s does not exist yet", annotation)
+		}
+		return err
+	}, backoff.WithMaxRetries(backoff.NewConstantBackOff(1*time.Second), retries))
+	return err == nil
+}
+
+// parseAnnotation parses the cni-proxy-config annotation into an iptables.Config object.
+func parseAnnotation(pod corev1.Pod, annotation string) (iptables.Config, error) {
+	anno, ok := pod.Annotations[annotation]
+	if !ok {
+		return iptables.Config{}, fmt.Errorf("could not find %s annotation for %s pod", annotation, pod.Name)
+	}
+	cfg := iptables.Config{}
+	err := json.Unmarshal([]byte(anno), &cfg)
+	if err != nil {
+		return iptables.Config{}, fmt.Errorf("could not unmarshal %s annotation for %s pod", annotation, pod.Name)
+	}
+	return cfg, nil
+}
```

Comment:
I feel this function is doing two separate things that should be divided up. The first is read the annotation from the Pod and the second is convert the value from the annotation to a `iptables.Config`. I think it might be cleaner and easier to test if you split these two actions into different functions.